libbeat/reader/readjson/docker_json.go

```diff
@@ -144,7 +144,7 @@ func (p *DockerJSONReader) parseDockerJSONLog(message *reader.Message, msg *logL
 }
 
 func (p *DockerJSONReader) parseLine(message *reader.Message, msg *logLine) error {
-	if strings.HasPrefix(string(message.Content), "{") {
+	if !p.criflags || len(message.Content) > 0 && message.Content[0] == '{' {
```

Comment:
`criflags` is used to enable parsing CRI flags within the CRI format, mainly intended for backwards compatibility, but this can still be CRI with it set to false

Reply:
The assumption here is, if criflags are enabled, we are dealing with CRI log. Otherwise check the first byte if it is '{' to determine the log format. Although a much better solution would be to detect log format at the start of the file and process it as either json-file or CRI log without doing this test for every line in the log.

As for your initial question, I don't have any proper benchmarks comparing old code to new code. I do have micro benchmarks comparing specific changes. I didn't benchmark []byte to string and then to []byte casting, since I'm dealing with very large volume of multi-part messages, that's 2 extra memory allocations per every entry and it adds up very quickly. It saves on memory and GC.

Reply:
I would prefer to avoid overloading this parameter and use something different to force the expected format. I plan to enable CRI flags parsing in our default manifests, as most users will benefit from them, but it should still work with docker logs.

I suggest we add a new `force_format: docker-json | cri` parameter. It can also be done in a following PR

Reply:
This is a lot better option then making assumptions. I will update this code in the new few days.

Reply:
@exekias After thinking this through, I've kept the current way of auto-detecting. But I've added an option to force CRI log parsing. Also I've moved new line removal to OS specific files, as such the correct function will be chosen at compile time.